extensions/redis-cache/src/main/java/io/druid/client/cache/RedisCache.java

```diff
@@ -0,0 +1,315 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.client.cache;
+
+import com.google.common.base.Function;
+import com.google.common.base.Supplier;
+import com.google.common.base.Suppliers;
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Maps;
+import com.metamx.common.logger.Logger;
+import com.metamx.emitter.service.ServiceEmitter;
+import io.druid.collections.LoadBalancingPool;
+import io.druid.collections.ResourceHolder;
+import io.druid.collections.StupidResourceHolder;
+import net.jpountz.lz4.LZ4Exception;
+import redis.clients.jedis.Jedis;
+import redis.clients.jedis.ShardedJedis;
+import redis.clients.jedis.exceptions.JedisConnectionException;
+import redis.clients.jedis.exceptions.JedisException;
+import redis.clients.util.Hashing;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class RedisCache implements Cache
+{
+  private static final Logger log = new Logger(RedisCache.class);
+
+  private final Supplier<ResourceHolder<ShardedJedis>> supplier;
+  private final String prefix;
+  private final long expiration;
+  private final RedisTranscoder transcoder;
+  private final AtomicLong hitCount = new AtomicLong(0);
+  private final AtomicLong missCount = new AtomicLong(0);
+  private final AtomicLong timeoutCount = new AtomicLong(0);
+  private final AtomicLong errorCount = new AtomicLong(0);
+
+  public static Cache create(final RedisCacheConfig config)
+  {
+    final Supplier<ResourceHolder<ShardedJedis>> clientSupplier;
+
+    if (config.getPoolSize() > 1) {
+      clientSupplier = new LoadBalancingPool<>(
```

Comment:
I think Jedis implements it's own connection pooling, which might be more efficient than maintaining multiple Jedis clients, so we might want to use that instead of our poor man's pooling we had to use for memcached.